### PR TITLE
NN-1953 whereabouts-api table changes

### PR DESCRIPTION
**Pull request by JamieDraperUK**
- add audit information
 - add case note id

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapper.java [hunk 1]
```diff
@@ -17,48 +16,45 @@
...
+        eventOutcomes.put(AbsentReason.UnacceptableAbsence, new EventOutcome("UNACAB", null));
 
-    private void addPaidAttendanceMapping(final String eventOutcome, final String performance) {
-        codeMap.put(getKey(null, true, true), new EventOutcome(eventOutcome, performance));
     }
 
-    private void addUnPaidAbsenceMapping(final AbsentReason reason, final String eventOutCome) {
-        codeMap.put(getKey(reason, false, false), new EventOutcome(eventOutCome, null));
-    }
+    EventOutcome getEventOutcome(final AbsentReason reason, final boolean attended, final boolean paid) {
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapper.java]: I'm planning on catching these exceptions in the controller, mapping them into 400 bad requests.

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/Attendance.java [hunk 2]
```diff
@@ -5,16 +5,23 @@
...
 import javax.validation.constraints.NotNull;
 import java.time.LocalDate;
+import java.time.LocalDateTime;
 
 @Entity
 @Table(name = "OFFENDER_ATTENDANCE")
 @Data()
 @NoArgsConstructor
 @AllArgsConstructor
+@EntityListeners(AuditingEntityListener.class)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/Attendance.java]: didn't realise you could do this :grin:

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/repository/AttendanceRepository.java [hunk 3]
```diff
@@ -12,4 +12,6 @@
 public interface AttendanceRepository extends CrudRepository<Attendance, Long> {
     Set<Attendance> findByPrisonIdAndEventLocationIdAndEventDateAndPeriod(String prisonId, Long eventLocationId,
                                                                           LocalDate date, TimePeriod period);
+
+    Attendance findById(long Id);
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/repository/AttendanceRepository.java]: why is this needed?

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/Attendance.java [hunk 4]
```diff
@@ -49,4 +56,22 @@
...
+
+    private long caseNoteId;
+
+    @CreatedDate
+    @Column(name = "CREATE_DATETIME", nullable = false)
+    private LocalDateTime creationDateTime;
+
+    @CreatedBy
+    @Column(name = "CREATE_USER_ID", nullable = false)
+    private String createUserId;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/Attendance.java]: not a great fan here - why is one `create` and the other `creation`?

File: src/main/resources/application-postgres.properties [hunk 5]
```diff
@@ -1,4 +1,4 @@
-spring.datasource.url=jdbc:postgresql://${DATABASE_ENDPOINT}/${DATABASE_NAME}
+spring.datasource.url=jdbc:postgresql://${DATABASE_ENDPOINT}/${DATABASE_NAME}?sslmode=verify-full
```
Comment (petergphillips) [inline src/main/resources/application-postgres.properties]: this looks naff - I would have expected a `DATASOURCE_URL` environment variable then these sort of production settings don't need to be here

File: src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/integration/AttendanceIntegrationTest.kt [hunk 6]
```diff
@@ -165,28 +168,28 @@ class AttendanceIntegrationTest : IntegrationTest () {
...
-                        .comments("hello world")
-                        .attended(false)
-                        .paid(false)
-                        .bookingId(1)
-                        .build()
-        )
+
+        val attendance = result.firstOrNull()!!
+
+        assertThat(attendance.id).isEqualTo(attendanceId)
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/integration/AttendanceIntegrationTest.kt]: I preferred the other version!  Can exclude individual fields with assertj...

File: src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceServiceTest.kt [hunk 7]
```diff
@@ -233,7 +259,10 @@ class AttendanceServiceTest {
 
     @Test
     fun `should record unpaid absence as 'Refused'`() {
-        val service = AttendanceService(attendanceRepository, nomisService)
+
+        `when`(nomisService.postCaseNote(
+                ArgumentMatchers.anyLong(), ArgumentMatchers.anyString(), ArgumentMatchers.anyString(),
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceServiceTest.kt]: best to import static on `ArgumentMatchers`

File: src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapperTest.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test
+    fun `should throw an exception when a unpaid reason is used for a paid attendance`() {
+        val unpaidReasons = AbsentReason.getUnpaidReasons()
+
+        unpaidReasons.forEach {
+            val reason = it
+            try {
+                eventOutcomeMapper.getEventOutcome(reason, false, true)
+                triggerFail()
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapperTest.kt]: horrible :grin: 
best to use `assertThatThrownBy` instead

File: src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapperTest.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        try {
+            eventOutcomeMapper.getEventOutcome(null, false, true)
+            triggerFail()
+        } catch (e: IllegalArgumentException) {
+            assertThat(e.message).isEqualTo("An absent reason was not supplied")
+        }
+    }
+
+    fun triggerFail() {
+        fail("should of thrown an IllegalArgumentException")
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/whereabouts/services/NomisEventOutcomeMapperTest.kt]: eugh
